src/common/snippets/src/lowered/pass/mark_invariant_shape_path.cpp

```diff
@@ -0,0 +1,124 @@
+           ov::op::util::is_binary_elementwise_logical(node) ||
+           ov::is_type<ov::op::v1::Select>(node) ||
+           ov::is_type<ov::snippets::op::Fill>(node) ||
+           ov::is_type<ov::snippets::op::ConvertTruncation>(node) ||
+           ov::is_type<ov::snippets::op::ConvertSaturation>(node);
+}
+static bool is_scalar_op(const ExpressionPtr& expr) {
+    const auto& node = expr->get_node();
+    return ov::is_type<ov::snippets::op::HorizonMax>(node) ||
+           ov::is_type<ov::snippets::op::HorizonSum>(node) ||
+           ov::is_type<ov::snippets::op::VectorBuffer>(node) ||
+           ov::is_type<ov::snippets::op::BroadcastMove>(node);
+}
+}  // namespace
+
+size_t MarkInvariantShapePath::getInvariantPortShapePath(const ExpressionPort& port) {
+    auto& rt = get_rt_info(port);
+    const auto rinfo = rt.find("InvariantShapePath");
+    OPENVINO_ASSERT(rinfo != rt.end(), "Invariant path for this expression port has not been marked!");
+    return rinfo->second.as<size_t>();
+}
+
+void MarkInvariantShapePath::SetInvariantPortShapePath(const ExpressionPort& port, size_t value) {
+    auto& rt = get_rt_info(port);
+    rt["InvariantShapePath"] = value;
+}
+
+ov::RTMap& MarkInvariantShapePath::get_rt_info(const ExpressionPort& port) {
+    const auto& node = port.get_expr()->get_node();
+    const auto port_idx = port.get_index();
+    const auto is_input = port.get_type() == ExpressionPort::Input;
+    OPENVINO_ASSERT((is_input && (port_idx < node->get_input_size())) || (!is_input && (port_idx < node->get_output_size())),
+                    "Node has incompatible port count with the expression");
+    return is_input ? node->input(port_idx).get_rt_info() : node->output(port_idx).get_rt_info();
+}
+
+bool MarkInvariantShapePath::run(lowered::LinearIR& linear_ir, lowered::LinearIR::constExprIt begin, lowered::LinearIR::constExprIt end) {
+    OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::MarkInvariantShapePath");
+
+    size_t color_path = 0;
+
+    auto merge_paths = [&color_path](size_t lhs, size_t rhs) {
+        if (lhs == rhs || rhs == NOT_AFFECTING_PATH) return lhs;
+        if (lhs == NOT_AFFECTING_PATH) return rhs;
+        return ++color_path;
+    };
+
+    for (auto expr_it = begin; expr_it != end; ++expr_it) {
+        const auto& expr = *expr_it;
+        if (ov::is_type<ov::snippets::op::LoopBase>(expr->get_node()))
+            continue;
+
+        for (size_t out_idx = 0; out_idx < expr->get_output_count(); ++out_idx) {
+            size_t current_color_path;
+            if (is_elementwise_op(expr)) {
+                current_color_path = NOT_AFFECTING_PATH;
+                for (size_t in_idx = 0; in_idx < expr->get_input_count(); ++in_idx) {
+                    const auto input_path = getInvariantPortShapePath(expr->get_input_port(in_idx));
+                    current_color_path = merge_paths(current_color_path, input_path);
+                }
```

Comment:
The main risk with `ov::Sybmol` is that it would likely be too slow to use in the whole shape-inference pipeline, so we should probably think about alternatives, and it could be one of them.
Let's just keep it in mind and discuss offline later.